### PR TITLE
[15.0][ADD] Purchase Order Merge

### PR DESCRIPTION
**Pull request by BojanOD**
Merge Purchase Orders

- https://github.com/oca/purchase-workflow/pull/676

### CODE REVIEW

**Review by ivantodorovich — CHANGES_REQUESTED** [review]
Pleasse add some unit tests 🙏🏻

File: purchase_merge/wizard/purchase_merge.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+        po.message_post(body=body, subject=subject, content_subtype="plaintext")
+
+    def _merge(self, purchases, dst_purchase=None):
+        """private implementation of merge purchase
+        :param purchases : ids of purchase to merge
+        :param dst_purchase : record of destination purchase.order
+
+        # TODO: to improved by using
+        OCA/openupgradelib/openupgrade_merge_records.py
```
Comment (ivantodorovich) [inline purchase_merge/wizard/purchase_merge.py]:
```suggestion
```

Seems like it's already implemented

File: purchase_merge/wizard/purchase_merge.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+                start_msg,
+                _("payment terms"),
+                ", ".join([str(x) for x in payment_term_names]),
+            )
+        if len(fiscal_position_names) > 1:
+            error_msg += "{} {}: {}\n".format(
+                start_msg,
+                _("fiscal positions"),
+                ", ".join([str(x) for x in fiscal_position_names]),
+            )
```
Comment (ivantodorovich) [inline purchase_merge/wizard/purchase_merge.py:103]: This `start_msg` / `error_msg` hack might work on english, but some translations may require a different gramatical structure. So this string concatenation won't work for them. Plus, it'll be a nightmare to translate as separated strings.

I recommend the not lazy approach and translate the full string for each case.
Also, compare records .. not record names.

I'd do something like this:

```py
error_messages = []

currencies = purchase_orders.currency_id
if len(currencies) > 1:
    error_messages.append(
        _(
            "You can't merge purchase orders with different currencies: %s",
            ", ".join(currencies.mapped("name"))
        )
    )

picking_types = purchase_orders.picking_type_id
if len(picking_types) > 1:
    error_messages.append(
        _(
            "You can't merge purchase orders with different picking types: %s",
            ", ".join(picking_types.mapped("name"))
        )
    )

# ... all the other checks ...

if error_messages:
    raise UserError("\n".join(error_messages))
```

File: purchase_merge/readme/ROADMAP.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+* it could be improved by using [https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade_merge_records.py](https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade_merge_records.py)
```
Comment (ivantodorovich) [inline purchase_merge/readme/ROADMAP.rst]: Remove this file. It's already implemented

**Review by ivantodorovich — CHANGES_REQUESTED** [review]

**Comment by imlopes** [comment]
@BojanOD 
Can you add the security files for the access to the wizard ?
https://www.odoo.com/pt_BR/forum/ajuda-1/why-access-rights-is-must-for-transient-model-in-odoo-14-181312

Without the security file, only Super User can access the wizard to merge.

File: purchase_merge/security/ir.model.access.csv [hunk 4]
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_purchase_merge_automatic_wizard_user,access_purchase_merge_automatic_wizard_user,model_access_purchase_merge_automatic_wizard,base.group_user,1,1,1,1
```
Comment (imlopes) [inline purchase_merge/security/ir.model.access.csv]: @BojanOD 

Any reason to use base.group_user instead Purchase User group ?

Comment (BojanOD) [reply purchase_merge/security/ir.model.access.csv]: @imlopes 

You are suggesting that I should use `purchase.group_purchase_user` instead?

Comment (imlopes) [reply purchase_merge/security/ir.model.access.csv]: For security reason, I think it's better. Even if a user without purchasing rights is prevented from seeing the wizard menu (merge), it is better to add the purchasing group..

**Review by imlopes — APPROVED** [review]
LGTM

**Review by ivantodorovich — APPROVED** [review]

**Comment by yankinmax** [comment]
@rvalyi @simahawk can we get it merged pls?

File: purchase_merge/wizard/purchase_merge.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        """
+        if len(purchases) < 2:
+            return
+        record_ids = purchases - dst_purchase
+        openupgrade_merge_records.merge_records(
+            env=self.env,
+            model_name=self._name,
+            record_ids=record_ids.ids,
+            target_record_id=dst_purchase.id,
+        )
```
Comment (quoc-pn) [inline purchase_merge/wizard/purchase_merge.py:183]: Hi @BojanOD,
This module is working great—thanks for the work on it.

I do have a question regarding the logic. It looks incorrect to me.
The model_name parameter refers to a wizard model (purchase.merge.automatic.wizard), but we're trying to merge purchase.order records. Both record_ids and target_record_ids belong to purchase.order.

With the current setup, the logic always returns None.
I’m testing this on PR #2544.

Could you take a look?

**Pull request merged**